src/persist-txn/src/txn_cache.rs

```diff
@@ -996,127 +996,118 @@ mod tests {
-        // range of time once you know of a later write that advances the
-        // physical upper.
-        // assert_eq!(cache.data_listen_next(&d0, 3), ReadDataTo(7));
-        // assert_eq!(cache.data_listen_next(&d0, 4), ReadDataTo(7));
-        assert_eq!(cache.data_listen_next(&d0, 5), ReadDataTo(7));
-        assert_eq!(cache.data_listen_next(&d0, 6), ReadDataTo(7));
+        // This attempts to exercise all the various interesting edge cases of
+        // data_snapshot and data_listen_subscribe using the following sequence
+        // of events:
+        //
+        // - Registrations at: [2,8], [12,13]
+        // - Direct writes at: 1, 10
+        // - Writes via txns at: 4, 5, 7
+        let mut c = TxnsCacheState::new(ShardId::new(), 0, None);
+
+        // empty
+        assert_eq!(c.progress_exclusive, 0);
+        assert!(std::panic::catch_unwind(|| c.data_snapshot(d0, 0)).is_err());
+        assert_eq!(c.data_listen_next(&d0, 0), WaitForTxnsProgress);
+
+        // ts 0 (never registered)
+        testcase(&mut c, 0, d0, ds(None, 0, 1), ReadDataTo(1));
+
+        // ts 1 (direct write)
+        testcase(&mut c, 1, d0, ds(None, 1, 2), ReadDataTo(2));
+
+        // ts 2 (register)
+        c.push_register(d0, 2, 1);
+        testcase(&mut c, 2, d0, ds(None, 2, 3), EmitLogicalProgress(3));
+
+        // ts 3 (registered, not written)
+        testcase(&mut c, 3, d0, ds(None, 3, 4), EmitLogicalProgress(4));
+
+        // ts 4 (written via txns)
+        c.push_append(d0, vec![4], 4, 1);
+        testcase(&mut c, 4, d0, ds(Some(4), 4, 5), ReadDataTo(5));
+
+        // ts 5 (written via txns, write at preceding ts)
+        c.push_append(d0, vec![5], 5, 1);
+        testcase(&mut c, 5, d0, ds(Some(5), 5, 6), ReadDataTo(6));
+
+        // ts 6 (registered, not written, write at preceding ts)
+        testcase(&mut c, 6, d0, ds(Some(5), 6, 7), EmitLogicalProgress(7));
+
+        // ts 7 (written via txns, write at non-preceding ts)
+        c.push_append(d0, vec![7], 7, 1);
+        testcase(&mut c, 7, d0, ds(Some(7), 7, 8), ReadDataTo(8));
+
+        // ts 8 (forget). Forget guarantees that all writes are applied, so no
+        // latest_write. It also currently is inclusive, so we can emit logical
+        // progress, though we might want to make the interval half-open?
+        c.push_register(d0, 2, -1);
+        testcase(&mut c, 8, d0, ds(None, 8, 9), EmitLogicalProgress(9));
+
+        // ts 9 (not registered, not written). This ReadDataTo would block until
+        // the write happens at ts 10.
+        testcase(&mut c, 9, d0, ds(None, 9, 10), ReadDataTo(10));
+
+        // ts 10 (written directly)
+        testcase(&mut c, 10, d0, ds(None, 10, 11), ReadDataTo(11));
+
+        // ts 11 (not registered, not written) This ReadDataTo would block until
+        // the register happens at 12.
+        testcase(&mut c, 11, d0, ds(None, 11, 12), ReadDataTo(12));
```

Comment:
I'm having a little bit of trouble figuring out the difference between all of these. We don't actually write to anything in any of them.

Reply:
Backing up a step, what I'm trying to solve here:

- For a table deletion, we forget the shard at some timestamp, removing it from the txns set
- We then advance the upper of the shard to empty antichain (this one of the necessary conditions to "delete" a shard)
- So, I would sort of expect that the txns_operator needs to continue to make progress even for shards that are no longer registered to the txns set.

- Similarly, because of the migration story, a shard might have been written to before it was registered. We probably want to ensure the txns_progress operator handles this situation correctly as well

- Finally, the unit test that registers and forgets stuff while doing direct and txns writes shook out a lot of bugs, so seems nice to support arbitrary interleavings here just so we can keep that test around

So, then, what are the expected semantics of txns_progress? I think my answer is that at any point it needs to be able to advance its own progress to`max(physical data shard upper, latest ts that the shard is registered in txns)`, the latter tracking the txns shard upper when the shard is registered.

Circling back to your question, this specific case tests that the shard is unregistered, it is not written to, and the txns shard advances. Basically, what happens is that we read what got written to the physical data shard at 11 (by reading to exclusive 12). If the next event in the data shard's timeline is a direct write, this read will resolve whenever that direct write happens. If the next event in the data shard's timeline is a register, then your PR will cause the data shard's upper to physically advance at that point, resolving that read.

Reply:
So the comments like `not written` and `written directly` are just hypothetical. From our point of view we just care that from `[9, 12)` we know that the data shard is not registered with the txn shard so readers need to consult the shard directly?

Reply:
> Curious if there's something I could write here that would have made it more obvious when you read it the first time?

Honestly, I'm not sure. It's probably fine as is.